File: python/279_PerfectSquares.py
from unittest import TestCase
# https://leetcode.com/problems/perfect-squares/
import math, collections


class PerfectSquares(TestCase):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        squares = [i*i for i in range(1, int(n ** 0.5)+1)]
        cnt = [n]*(n+1)
        cnt[0] = 0
        for x in range(n):
            for s in squares:
                v = x+s
                if v > n: break
                cnt[v] = min(cnt[v], cnt[x]+1)
        return cnt[n]

    # # bfs
    # def numSquares(self, n):
    #     squares = [i*i for i in range(1, int(math.sqrt(n))+1)]
    #     que = collections.deque([(0, 0, 0)])
    #     while que:
    #         v, c, i = que.popleft()
    #         for j in range(i, len(squares)):
    #             nxt = v + squares[j]
    #             if nxt == n: return c+1
    #             if nxt > n: break
    #             que.append((nxt, c+1, j))
    #     return -1
    #
    # # theory number O(sqrt(n)): must <=4
    # def numSquares(self, n):
    #     sq = int(math.sqrt(n))+1
    #     for a in range(sq):
    #         for b in range(a, sq):
    #             if a*a + 2*b*b > n: break
    #             c = int(math.sqrt(n - a*a - b*b))
    #             if a*a + b*b + c*c == n:
    #                 return 1 + bool(a) + bool(b)
    #     return 4

    def test1(self):
        self.assertEqual(3, self.numSquares(12))

    def test2(self):
        self.assertEqual(2, self.numSquares(13))

    def test3(self):
        self.assertEqual(2, self.numSquares(7))

File: kotlin/src/main/kotlin/279.PerfectSquares.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/perfect-squares/


class PerfectSquares{
  fun numSquares(n: Int): Int {
    val cnt = IntArray(n+1){ n }
    val squares = mutableListOf(1)
    for(x in 2..n) {
      val s = x*x
      if (s > n) break
      squares.add(s)
      cnt[s] = 1
    }

    for(x in 2..n)
      for(s in squares)
        if(x+s > n || s > x)
          break
        else
          cnt[x+s] = minOf(cnt[x+s], cnt[x]+1)

    return cnt[n]
  }

  //bfs
  fun numSquares1(n: Int): Int {
    val squares = mutableListOf(1)
    for(x in 2..n) {
      val s = x * x
      if (s > n) break
      squares.add(s)
    }
    val set = mutableSetOf(0)
    var layer = mutableSetOf(0)
    var cnt = 0
    while(cnt <= n){
      cnt ++
      val next = mutableSetOf<Int>()
      for(v in layer)
        for(s in squares) {
          val x = v+s
          if (x == n) return cnt
          else if (x < n && !set.contains(x)) {
            set.add(x)
            next.add(x)
          }
        }
      layer = next
    }
    return -1
  }

  @Test
  fun test1(){
    assertEquals(3, numSquares(12))
  }

  @Test
  fun test2(){
    assertEquals(2, numSquares(13))
  }
}